data/abi/abi_encode.go

```diff
@@ -0,0 +1,444 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package abi
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/big"
+)
+
+// arrayToTuple casts an array-like ABI Value into an ABI Value of Tuple type.
+// This is used in both ABI Encoding and Decoding.
+func (v Value) arrayToTuple() (Value, error) {
+	var childT []Type
+	var valueArr []Value
+
+	switch v.ABIType.enumIndex {
+	case String:
+		strValue, err := v.GetString()
+		if err != nil {
+			return Value{}, err
+		}
+		strByte := []byte(strValue)
+
+		childT = make([]Type, len(strByte))
+		valueArr = make([]Value, len(strByte))
+
+		for i := 0; i < len(strByte); i++ {
+			childT[i] = MakeByteType()
+			valueArr[i] = MakeByte(strByte[i])
+		}
+	case Address:
+		addr, err := v.GetAddress()
+		if err != nil {
+			return Value{}, err
+		}
+
+		childT = make([]Type, 32)
```

Comment:
there was a func like `ByteLen()` - use it (or a named constant) to minimize numeric literals usage